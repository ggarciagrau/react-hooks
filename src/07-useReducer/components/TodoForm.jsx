import { useState } from "react";
import { useForm } from "../../hooks";

export const TodoForm = () => {

    const { description, update, reset } = useForm({
        description: ""
    })

    const onSubmit = (ev) => {
        ev.preventDefault();

        const { target: { description }} = ev;

        if (!description) return;

        reset();
    }

    return (
        <form onSubmit={onSubmit}>
            <input
                type="text"
                placeholder="New todo"
                className="form-control"
                value={description}
                onChange={({ target: { value: description } }) => update({ ...description })}
            />

            <button
                type="submit"
                className="btn btn-outline-primary mt-2"
            >
                Add
            </button>
        </form>
    );
}