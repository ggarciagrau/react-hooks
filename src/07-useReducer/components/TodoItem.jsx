
export const TodoItem = ({ id, description, checked, handleRemoveTodo, toggleTodo }) => {

    return (
        <li className="list-group-item d-flex justify-content-between">
            <span className="align-self-center">{description}</span>
            <input type="checkbox" checked={checked} onChange={() => toggleTodo(id)} />
            <button
                className="btn btn-danger"
                onClick={() => handleRemoveTodo(id)}
            >
                Delete
            </button>
        </li>
    );
}