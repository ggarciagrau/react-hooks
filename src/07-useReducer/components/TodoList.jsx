
export const TodoList = ({ children }) => {

    return (
        <ul className="list-group">
            {children}
        </ul>

    );
}