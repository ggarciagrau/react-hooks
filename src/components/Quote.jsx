import propTypes from "prop-types";

export const Quote = ({ quote, author }) => {

    return (
        <blockquote className="blockquote text-end">
            <p className="mb-1">{quote}</p>
            <footer className="blockquote-footer">{author}</footer>
        </blockquote>
    );
}

Quote.propTypes = {
    quote: propTypes.string.isRequired,
    author: propTypes.string.isRequired
}