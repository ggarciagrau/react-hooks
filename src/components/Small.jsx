import { memo } from 'react';
import propTypes from 'prop-types';

export const Small = memo(({ number }) => {

    console.log('Small component called!');

    return (
        <small>{number}</small>
    );
});

Small.propTypes = {
    number: propTypes.number.isRequired
}