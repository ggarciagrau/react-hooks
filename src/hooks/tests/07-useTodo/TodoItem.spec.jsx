import { fireEvent, act, render } from "@testing-library/react";
import { TodoItem } from "../../../07-useReducer/components/TodoItem";

const id = 55;
let checked = false;

const mockToggleTodo = jest.fn();

mockToggleTodo.mockImplementation(() => {
  checked = !checked;
});

const mockHandleRemoveTodo = jest.fn();

// id, description, checked, handleRemoveTodo, toggleTodo

describe("Tests for TodoItem", () => {
  it("Should render by default", () => {
    const rendered = render(
      <TodoItem
        id={id}
        description={"Product"}
        checked={checked}
        handleRemoveTodo={mockHandleRemoveTodo}
        toggleTodo={mockToggleTodo}
      />
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it("Should remove the todo", () => {
    const rendered = render(
      <TodoItem
        id={id}
        description={"Product"}
        checked={checked}
        handleRemoveTodo={mockHandleRemoveTodo}
        toggleTodo={mockToggleTodo}
      />
    );

    const removeButton = rendered.getByText("Delete");

    act(() => {
      fireEvent.click(removeButton);
    });

    expect(mockHandleRemoveTodo).toHaveBeenCalledWith(id);
  });

  it("Should toggle the todo", async () => {
    const rendered = render(
      <TodoItem
        id={id}
        description={"Product"}
        checked={checked}
        handleRemoveTodo={mockHandleRemoveTodo}
        toggleTodo={mockToggleTodo}
      />
    );

    const checkbox = rendered.getByRole("checkbox");
    act(() => {
      fireEvent.click(checkbox);
    });

    rendered.rerender(
        <TodoItem
          id={id}
          description={"Product"}
          checked={checked}
          handleRemoveTodo={mockHandleRemoveTodo}
          toggleTodo={mockToggleTodo}
        />
      );

    expect(mockToggleTodo).toHaveBeenCalled();
    expect(checkbox.checked).toBeTruthy();
  });
});
