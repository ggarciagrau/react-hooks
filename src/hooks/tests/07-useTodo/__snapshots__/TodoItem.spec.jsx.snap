// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TodoItem Should render by default 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      class="align-self-center"
    >
      Product
    </span>
    <input
      type="checkbox"
    />
    <button
      class="btn btn-danger"
    >
      Delete
    </button>
  </li>
</div>
`;
